src/TreeSelect.tsx

```diff
@@ -56,8 +56,8 @@ export default class TreeSelect extends React.Component<Props, State> {
     treeVisible: false,
   };
 
-  public componentDidUpdate(_: Props, prevState: State) {
-    if (this.state.treeVisible && !prevState.treeVisible && this.filterInput)
+  public componentDidUpdate() {
+    if (this.state.treeVisible && this.filterInput)
```

Comment:
A standard pattern in `componentDidUpdate` is `if (prevProp !== newProp) { doSmth(); }`. This is because technically there is no clear guarantee how few times `componentDidUpdate` will be called. Theoretically, [the docs](https://facebook.github.io/react/docs/react-component.html#componentdidupdate) don’t  specify what an “update” is. More practically, as soon as someone extends the component, this might correctly fire many times a second. Whether I am tracking a mouse position, keyboard input, or just the length of a text field – everything can go into the state. As a result, whichever imperative, side-effect driven functionality you put here, will be called too often.

In this specific case, it will steal focus from anything else all the time due to the now unrestricted `.focus()` call. Instead, consider explicitly describing just which change in props/state should trigger the behavior. Here it’s probably `if (this.state.filter !== prevState.filter)`.

Reply:
Great points. Updated.